packages/alignments/src/CramAdapter/CramAdapter.ts

```diff
@@ -18,200 +25,218 @@ interface Header {
   nameToId?: Record<string, number>
 }
 
-export default class CramAdapter extends BaseAdapter {
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  private cram: any
+export default (pluginManager: PluginManager) => {
+  class CramAdapter extends BaseFeatureDataAdapter {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    private cram: any
 
-  private sequenceAdapter: BaseAdapter
+    private sequenceAdapter: BaseFeatureDataAdapter
 
-  private samHeader: Header = {}
+    private samHeader: Header = {}
 
-  // maps a refname to an id
-  private seqIdToRefName: string[] | undefined
+    // maps a refname to an id
+    private seqIdToRefName: string[] | undefined
 
-  // maps a seqId to original refname, passed specially to render args, to a seqid
-  private seqIdToOriginalRefName: string[] = []
+    // maps a seqId to original refname, passed specially to render args, to a seqid
+    private seqIdToOriginalRefName: string[] = []
 
-  public static capabilities = ['getFeatures', 'getRefNames']
+    public constructor(
+      config: AnyConfigurationModel,
+      getSubAdapter: getSubAdapterType,
+    ) {
+      super(config)
 
-  public constructor(config: {
-    cramLocation: IFileLocation
-    craiLocation: IFileLocation
-    sequenceAdapter: BaseAdapter
-    fetchSizeLimit?: number
-  }) {
-    super(config)
-    const { cramLocation, craiLocation, sequenceAdapter } = config
-    if (!cramLocation) {
-      throw new Error('missing cramLocation argument')
-    }
-    if (!craiLocation) {
-      throw new Error('missing craiLocation argument')
+      const cramLocation = readConfObject(config, 'cramLocation')
+      const craiLocation = readConfObject(config, 'craiLocation')
+      if (!cramLocation) {
+        throw new Error('missing cramLocation argument')
+      }
+      if (!craiLocation) {
+        throw new Error('missing craiLocation argument')
+      }
+      this.cram = new IndexedCramFile({
+        cramFilehandle: openLocation(cramLocation),
+        index: new CraiIndex({ filehandle: openLocation(craiLocation) }),
+        seqFetch: this.seqFetch.bind(this),
+        checkSequenceMD5: false,
+        fetchSizeLimit: config.fetchSizeLimit || 600000000,
+      })
+
+      // instantiate the sequence adapter
+      const sequenceAdapterType = readConfObject(config, [
+        'sequenceAdapter',
+        'type',
+      ])
+      const sequenceAdapterConfig = readConfObject(config, 'sequenceAdapter')
+      const sequenceAdapter = getSubAdapter(
+        sequenceAdapterType,
+        sequenceAdapterConfig,
+      ).dataAdapter
+      if (sequenceAdapter instanceof BaseFeatureDataAdapter) {
```

Comment:
This subadapter concept overall looks quite nice. Any reason why this instanceof check is needed though? instanceof is quite inflexible if someone did want to reimplement the base class themselves

Reply:
there are now 2 data adapter base types: one for refname aliases, and one for everything else. currently the basefeaturedataadapter is a thing that has to be inherited from but i suppose we could make it an abstract interface down the road